src/shims/env.rs

```diff
@@ -127,18 +128,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         // If we cannot get the current directory, we return null
         match env::current_dir() {
             Ok(cwd) => {
-                // It is not clear what happens with non-utf8 paths here
-                let mut bytes = cwd.display().to_string().into_bytes();
-                // If `size` is smaller or equal than the `bytes.len()`, writing `bytes` plus the
-                // required null terminator to memory using the `buf` pointer would cause an
-                // overflow. The desired behavior in this case is to return null.
-                if (bytes.len() as u64) < size {
-                    // We add a `/0` terminator
-                    bytes.push(0);
-                    // This is ok because the buffer was strictly larger than `bytes`, so after
-                    // adding the null terminator, the buffer size is larger or equal to
-                    // `bytes.len()`, meaning that `bytes` actually fit inside tbe buffer.
-                    this.memory.write_bytes(buf, bytes.iter().copied())?;
+                if this.write_os_str_to_c_string(&OsString::from(cwd), buf, size).is_ok() {
```

Comment:
That function can fail for various reasons though -- like, the ptr could be dangling. That would be UB.

Please separate "allowed" errors (buffer size too small) from `InterpError`. Probably `write_os_str_to_c_string` should return `InterpResult<'tcx, bool>` or so to indicate if writing was successful.